Refactor SUMTHIN Implementation

bska opened the pull request:
Report steps do not influence the `SUMTHIN` intervals and the special value
```
SUMTHIN
0.0 /
```
turns off the "thinning" process.  This refactoring introduces a new helper variable, `sumthin_active_`, which caches whether `SUMTHIN` is active (i.e., exists and has strictly positive value).

While here, also ensure that the ministep IDs have non-unit steps when `SUMTHIN` ends up skipping an output event.

bska commented:
jenkins build this please

joakim-hove commented:
The sumthin implementasjon is manager by scheduledtate - why do you nerd to add the new starte variabler to the summary? 

Without have lookef om detail om this - or knowing the problem you intens to selve I am a bit skeptisk!

bska commented:
> why do you nerd to add the new starte variabler to the summary?

The current implementation doesn't work.  This PR does.

bska commented:
jenkins build this please

OPMUSER commented:
My tests fail for SUMTHIN, see attached.

[SUMTHIN.zip](https://github.com/OPM/opm-common/files/7090857/SUMTHIN.zip)

I bet you are really happy that Joakim showed me how to compile from source :-)

bska commented:
> My tests fail for SUMTHIN, see attached.

Thanks for the report.  To be honest I'm a little bit puzzled by this case.  Your `SUMTHIN` request is
```
SUMTHIN
  91.3125 /
```
but your report steps are all around 30 days.  It's not clear to me what you expect to happen.  Unless I'm completely misreading my reference material&mdash;which is admittedly entirely possible&mdash;we're supposed to always output summary vectors whenever there's a report step whereas the `SUMTHIN` request only applies to timesteps outside of the report steps.

I moved `SUMTHIN` to the `SCHEDULE` section because I expect we have the same issue as with [`RPTONLY`](https://github.com/OPM/opm-common/pull/2597#issuecomment-910067643) in terms of recognising `SUMTHIN` in the `SUMMARY` section.  I also added
```
TUNING
 1* 5 /
 /
 /
```
to limit the timestep sizes (<= 5 days) and then I got summary output at plausible times.  At least given the restriction that we always output summary at report steps.  If `SUMTHIN` is supposed to apply to report steps too then we'll need change this implementation.

OPMUSER commented:
@bska Okay, the results are in and there are minor differences:

![image](https://user-images.githubusercontent.com/17958927/131948769-23b3e3d4-f6b4-4688-8eb2-54c381d073b2.png)

**I'm okay with this and will close the ticket if you are okay with this.**

OPMUSER commented:
The above is without the TUNING keyword, I'll run another case with the TUNING keyword active.

OPMUSER commented:
Okay, and the results are in for the second and last time. 

Basically they are similar but not the same, so I think this is acceptable given the difference in time stepping etc. and bearing in mind that we are just wishing to reduce the number of entries in the SUMMARY file. We just need to fix the keyword not working in the SUMMARY section for the release. The SUMTHIN-01 case is with the TUNING keyword active and set to maximum TSTEP of 5 days.

![image](https://user-images.githubusercontent.com/17958927/132157514-f7bc4c7f-feb1-40ef-906d-17b75009d408.png)

joakim-hove commented:
>Agreed. I think I have an idea for how to do that without causing too much disruption to the existing system.

I think this can be achieved by initializing the first ScheduleState instance with a value talen from the Schedule member ScheduleStatic - there Are some other properties initialized this way.

bska commented:
> > Agreed. I think I have an idea for how to do that without causing too much disruption to the existing system.
> 
> I think this can be achieved by initializing the first ScheduleState instance with a value talen from the Schedule member ScheduleStatic - there Are some other properties initialized this way.

This was exactly what I had in mind.

bska commented:
> > > Agreed. I think I have an idea for how to do that without causing too much disruption to the existing system.
> > 
> > I think this can be achieved by initializing the first `ScheduleState` instance with a value taken from the `Schedule` member `ScheduleStatic` - there are some other properties initialized this way.
> 
> This was exactly what I had in mind.

This approach is implemented in #2761.
